File: mmupnp/src/main/java/net/mm2d/upnp/internal/thread/ExecutorFactory.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.thread;

import net.mm2d.upnp.TaskExecutor;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import javax.annotation.Nonnull;

class ExecutorFactory {
    private static final int PRIORITY_CALLBACK = Thread.NORM_PRIORITY;
    private static final int PRIORITY_IO = Thread.MIN_PRIORITY;
    private static final int PRIORITY_MANAGER = Thread.MIN_PRIORITY;
    private static final int PRIORITY_SERVER = Thread.MIN_PRIORITY + 1;

    @Nonnull
    static TaskExecutor createCallback() {
        final ThreadFactory factory = new ExecutorThreadFactory("cb-", PRIORITY_CALLBACK);
        final ExecutorService executor = Executors.newSingleThreadExecutor(factory);
        return new DefaultTaskExecutor(executor);
    }

    @Nonnull
    static TaskExecutor createIo() {
        return createIo(calculateMaximumPoolSize());
    }

    private static int calculateMaximumPoolSize() {
        return Math.max(2, Runtime.getRuntime().availableProcessors()) * 2;
    }

    @Nonnull
    static TaskExecutor createIo(final int maxThread) {
        final ThreadFactory factory = new ExecutorThreadFactory("io-", PRIORITY_IO);
        final ThreadWorkQueue queue = new ThreadWorkQueue();
        final ExecutorService executor = new ThreadPoolExecutor(0, maxThread, 15L, TimeUnit.SECONDS, queue, factory, queue);
        return new DefaultTaskExecutor(executor, true);
    }

    @Nonnull
    static TaskExecutor createManager() {
        final ExecutorService executor = createServiceExecutor("mg-", PRIORITY_MANAGER);
        return new DefaultTaskExecutor(executor);
    }

    @Nonnull
    static TaskExecutor createServer() {
        final ExecutorService executor = createServiceExecutor("sv-", PRIORITY_SERVER);
        return new DefaultTaskExecutor(executor, true);
    }

    @Nonnull
    private static ExecutorService createServiceExecutor(
            @Nonnull final String prefix,
            final int priority) {
        final ThreadFactory factory = new ExecutorThreadFactory(prefix, priority);
        return new ThreadPoolExecutor(0, Integer.MAX_VALUE,
                0L, TimeUnit.NANOSECONDS,
                new SynchronousQueue<>(), factory);
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/thread/ExecutorFactory.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.thread

import java.util.concurrent.*

internal object ExecutorFactory {
    private const val PRIORITY_CALLBACK = Thread.NORM_PRIORITY
    private const val PRIORITY_IO = Thread.MIN_PRIORITY
    private const val PRIORITY_MANAGER = Thread.MIN_PRIORITY
    private const val PRIORITY_SERVER = Thread.MIN_PRIORITY + 1
    private const val KEEP_ALIVE_SECOND = 15L

    fun callback(): ExecuteFunction {
        val factory = ExecutorThreadFactory("callback-", PRIORITY_CALLBACK)
        val executor = Executors.newSingleThreadExecutor(factory)
        return DefaultTaskExecutor(executor).toFunction()
    }

    fun io(maxThread: Int = maxPoolSize()): ExecuteFunction {
        val factory = ExecutorThreadFactory("io-", PRIORITY_IO)
        val queue = ThreadWorkQueue()
        val executor = ThreadPoolExecutor(0, maxThread, KEEP_ALIVE_SECOND, TimeUnit.SECONDS, queue, factory, queue)
        return DefaultTaskExecutor(executor, true).toFunction()
    }

    private fun maxPoolSize(): Int = maxOf(2, Runtime.getRuntime().availableProcessors()) * 2

    fun manager(): ExecuteFunction =
        DefaultTaskExecutor(serviceExecutor("mg-", PRIORITY_MANAGER)).toFunction()

    fun server(): ExecuteFunction =
        DefaultTaskExecutor(serviceExecutor("sv-", PRIORITY_SERVER), true).toFunction()

    private fun serviceExecutor(prefix: String, priority: Int): ExecutorService =
        ThreadPoolExecutor(
            0, Integer.MAX_VALUE,
            0L, TimeUnit.NANOSECONDS,
            SynchronousQueue(),
            ExecutorThreadFactory(prefix, priority)
        )
}
